client/src/Stores/MicrosoftLearn.store.ts

```diff
@@ -155,4 +233,130 @@ export class MicrosoftLearnStore extends ChildStore {
   private scoreRegex(testPhrase: string, exp: RegExp, score = 1): number {
     return exp.test(testPhrase) ? score : 0;
   }
-}
+
+    // O(|filtertype|)
+  private removeExtrasFromSelected = (type: FilterType) => {
+    if(type===FilterType.Product){
+    
+          let parentProducts = Array.from(this.productMap.keys());
+  
+          // CASE 1: check for parent products which are there in selectedFilter but not in displayFilter. Remove all of them from selectedFilter, along with all their children.
+          // CASE 2: check for individual product filters whose parent is not in selected filter, but it is, and the current state of displayFilter does not. Remove them from selectedFilter.
+  
+          let prevSelected = this.selectedFilters.get(type);
+          let newDisplay = this.displayFilters.get(type);
+          let removeProducts: string[] = []
+          parentProducts?.forEach(item=>{
+          if(newDisplay && prevSelected){
+            if(prevSelected.includes(item.id) && !newDisplay.includes(item.id))
+                  removeProducts.push(item.id);
+          }
+          })
+          let productIter = this.catalog?.products.values();
+          let _n = 0;
+          while(this.catalog?.products.size && _n<this.catalog?.products.size){
+            let p = productIter?.next().value;
+            if((removeProducts.includes(p.parentId) ||
+            (!prevSelected?.includes(p.parentId) && prevSelected?.includes(p.id) && !newDisplay?.includes(p.id)))){
+                removeProducts.push(p.id);

```

Comment:
Yes, I agree with this. In terms of multi-level hierarchy, we would have to consider how it renders on the UI. If say in the future we end up having 3/4 levels of hierarchy for a particular item, rendering it in the same view with the indentations may not look good from the UX perspective. One possibility could be- we handle all levels of hierarchy that come in from the catalog API, but at our front end side, we combine all the lower level hierarchies into one, i.e., from the user's perspective, we show him only two levels of products. How do you think about this?

Reply:
Hmm. So, the multi-level work comes with its own UI baggage as well. In that case, let's park it for now. Can we consider keeping our logic written in a manner that it could easily scale multi-level? For example, in this case, it seems like we'd require to completely re-think and re-write this piece of code to make it work multi-level.

Reply:
I went over the logic once again to see where and how the logic would need changes in case we encounter multi-hierarchy products in the future. I have added comments for the same in my latest commit. Considering that the `Product.dto` and `catalog?.product` are changed accordingly, we would need minimal changes with the existing logic. The above particular piece of code can also stay intact if we go by that approach. 

To elaborate upon the same a little, in particular:

- We can extend the `productHierarchicalMap` (`getProductHierarchicalMap`) by having keys for all those children as well which have further sub-items; for instance:

      P1
               .>C1
                        .>GC1
                        .>GC2

      P2
               .>C2
       
      The above structure would appear in the `productHierarchicalMap` (`productParentChildMap`) as:
      P1: [C1]
      P2: [C2]
      C1: [GC1, GC2]  

- Given we construct the above map, wrt the logic, we would need changes in
    - `setDisplayFilters`: add all ancestors instead of only the parents.
    - `removeExtraFilters`: this would stay intact.
    - `updateUri`: minor change to this as per comment in code.

- The UI would have to be changed accordingly as well depending on how we want to style it.
If the above suggestions seem fine, how do you propose- should we make these changes now only? The only problem I see with that is that we won't be able to test it all completely as of now.